### PR TITLE
rustdoc: Skip doc link resolution for non-exported items

### PR DESCRIPTION
**Pull request by petrochenkov**

### CODE REVIEW

**Comment by petrochenkov** [comment]
Are benchmarks run with `--document-private-items` by chance?

Anyway, this is a cheap optimization that can skip extra work in some cases, and it's done in other situations (`ResolveDocLinks::ExportedMetadata`), so it still makes sense to land it.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 1]
```diff
@@ -949,6 +924,43 @@ fn preprocessed_markdown_links(s: &str) -> Vec<PreprocessedMarkdownLink> {
 }
 
 impl LinkCollector<'_, '_> {
+    fn resolve_links(&mut self, item: &Item) {
+        if !self.cx.render_options.document_private
+            && let Some(def_id) = item.item_id.as_def_id()
+            && let Some(def_id) = def_id.as_local()
+            && !self.cx.tcx.effective_visibilities(()).is_exported(def_id)
```
Comment (jyn514) [inline src/librustdoc/passes/collect_intra_doc_links.rs:931]: Are you sure `is_exported` is the correct thing here, and not `is_reachable`? Can you add a test case for
```rust
mod private {
  /// [core::str::FromStr]
  pub struct Bar;
}

pub fn foo() -> private::Bar {
    private::Bar
}
```
and make sure it behaves the same both before and after this change?

**Review by jyn514 — APPROVED** [review]
The code looks good and the change makes sense to me (this is just a perf optimization, right, it shouldn't affect behavior?).

r=me with the suggested test case added.

**Comment by jyn514** [comment]
Actually hold on - all private items should already be stripped from the crate at this point (see https://github.com/rust-lang/rust/blob/0e8f0b03cd94d9611063daaa516af2ae3489c6b5/src/librustdoc/passes/strip_private.rs#L15-L17). I think that's why this isn't showing a perf improvement, it's not actually affecting rustdoc's behavior. What issue were you running into before you added the `!is_exported()` check to the intra-doc links pass?

**Comment by petrochenkov** [comment]
@jyn514 
>What issue were you running into before you added the `!is_exported()` check to the intra-doc links pass?

And ICE when documenting liballoc.
```
Documenting alloc v0.0.0 (C:\msys64\home\we\rust\library\alloc)
thread 'rustc' panicked at 'no resolutions for a doc link', compiler\rustc_metadata\src\rmeta\encoder.rs:2251:18
```
rustdoc still tries to request resolution for a link on some non-exported item.

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 1 above)
Comment (petrochenkov) [reply src/librustdoc/passes/collect_intra_doc_links.rs:905]: In both cases `Bar` is not documented and `Bar` in `pub fn foo() -> Bar` in the generated doc is not a link.

Rustdoc uses `is_reachable` in only one place when it generates json output, and I'm not sure why.

**Comment by petrochenkov** [comment]
Inner doc comment on the `alloc::vec::in_place_collect` module, which is indeed not exported (it's fully private).
```
[src\librustdoc\passes\collect_intra_doc_links.rs:377] path_str = "Vec"
[src\librustdoc\passes\collect_intra_doc_links.rs:377] ns = MacroNS
[src\librustdoc\passes\collect_intra_doc_links.rs:377] item_id = DefId(0:6408 ~ alloc[76f0]::vec::in_place_collect)
[src\librustdoc\passes\collect_intra_doc_links.rs:377] module_id = DefId(0:6408 ~ alloc[76f0]::vec::in_place_collect)
thread 'rustc' panicked at 'no resolutions for a doc link', compiler\rustc_metadata\src\rmeta\encoder.rs:2255:18
```

**Comment by jyn514** [comment]
> rustdoc still tries to request resolution for a link on some non-exported item.
> 
> It seems like it doesn't happen often due to the item stripping pass, so there's no effect on performance, but does happen.

This seems like a pre-existing bug I would like to investigate, but it doesn't have to block this PR.

@bors r+

**Comment by ehuss** [comment]
@bors retry

github incident

**Pull request merged**